rust/datafusion/src/physical_plan/limit.rs

```diff
@@ -236,23 +245,29 @@ impl Stream for LimitStream {
         mut self: Pin<&mut Self>,
         cx: &mut Context<'_>,
     ) -> Poll<Option<Self::Item>> {
-        self.input.poll_next_unpin(cx).map(|x| match x {
-            Some(Ok(batch)) => Ok(self.stream_limit(batch)).transpose(),
-            other => other,
-        })
+        match &mut self.input {
```

Comment:
Why not just compare `self.current_len == self.limit` and short-cirtcuit before polling the wrapped stream, instead of the Option plumbing?

Reply:
My thinking was that the option plumbing actually `drop`s the input, freeing its resources when the limit has been hit, rather than waiting for the execution to be complete.

Reply:
Fair enough

Reply:
Do we do this in other places too? Isn't a `SendableRecordBatchStream` a small struct?

Reply:
> Isn't a SendableRecordBatchStream a small struct?

It is a trait, so there are various things that implement it. Some, like the `ParquetStream` https://github.com/apache/arrow/blob/b6982df45dfd0cebf2b34deaec62801f87242bc5/rust/datafusion/src/physical_plan/parquet.rs#L876, could have substantial resources

Reply:
As far as I understand it now, it can consist of the whole tree of dependent streams. Probably still not a big resource hog but more than a few bytes.

Reply:
Imagine it is actually a subquery with a group by hash or join with a large hash table :) It may actually be hanging on to a substantial amount of memory I suspect

Reply:
Ah, yeah that's right!